BrainPortal/lib/s3_sdkv3_connection.rb

```diff
@@ -0,0 +1,346 @@
+    keyPath = dest_head.nil? ? trun_object_key : File.join(dest_head,trun_object_key).to_s
+    if srcStream.nil?
+      File.open(srcFile,'r') do |srcIO|
+        resp = @resource.client.put_object(body: srcIO,
+                                           bucket: @bucket_name,
+                                           key: keyPath)
+      end
+    else
+        resp = @resource.client.put_object(body: srcStream,
+                                           bucket: @bucket_name,
+                                           key: keyPath)
+    end
+  end
+
+  # Copies a directory to the bucket, will recursively create directory in the path
+  def copy_directory_to_bucket(srcDir, dest_head)
+    if object_exists?(dest_head)
+      Dir.glob("#{srcDir}/**/*").each do |x|
+        next if File.directory?(x)
+        sub_dir_mon = "#{File.dirname(srcDir)}/"
+        newDirName = "#{File.dirname(x).sub(sub_dir_mon,'')}"
+        new_dest_path = File.join(clean_starting_folder_path(dest_head).to_s,newDirName)
+        srcStream = nil
+        if File.symlink?(x)
+          symlink_filename = "#{x}#{@symlink_ending}"
+          srcStream = Tempfile.new()
+          srcStream.write(File.readlink(x))
+          srcStream.rewind
+          copy_file_to_bucket(symlink_filename,new_dest_path,srcStream)
+          srcStream.unlink
+          srcStream.close
+        else
+          copy_file_to_bucket(x,new_dest_path)
+        end
+      end
+    end
+  end
+
+  # Delete an object from the bucket
+  def delete_path_from_bucket(path)
+    path_clean = clean_starting_folder_path(path).to_s
+    if object_exists?(path_clean)
+      @resource.bucket(@bucket_name).objects({prefix: path_clean}).batch_delete!
+    end
+  end
+
+  # Renames an objects path
+  def rename_object(srcPath,destPath)
+    src_path_clean = clean_starting_folder_path(srcPath).to_s
+    dest_path_clean = clean_starting_folder_path(destPath).to_s
+    if object_exists?(src_path_clean)
+      src_path_w_bucket = "/#{File.join(@bucket_name,src_path_clean)}"
+      resp = @resource.client.copy_object(bucket:@bucket_name,
+                                          copy_source: src_path_w_bucket,
+                                          key: dest_path_clean)
+      delete_path_from_bucket(src_path_clean)
+    end
+  end
+
+  # Renames a Folder by renaming each object underneath
+  def rename_path(srcPath,destPath)
+    src_path_clean = clean_starting_folder_path(srcPath).to_s
+    dest_path_clean = clean_starting_folder_path(destPath).to_s
+
+    if object_exists?(src_path_clean)
+      listObjects = list_objects_long(src_path_clean)
+
+      listObjects.each do |x|
+        xStat = get_object_stats(x[:key])
+        next if xStat[:content_type] == 'application/x-directory'
+        mod_dest_path = x[:key].dup.sub! src_path_clean,dest_path_clean
+        rename_object(x[:key],mod_dest_path)
+      end
+    end
+  end
+
+  # List the buckets available
+  def list_buckets
```

Comment:
This method doesn't seem to be used by anything?

Reply:
Correct, it is included in the protocol library for completeness.